src/test/integ/highlight-grid.test.ts

```diff
@@ -0,0 +1,471 @@
+                    lineText: "👋",
+                    tabSize: 4,
+                },
+            ],
+            expectedRanges: [
+                {
+                    textType: "virtual" as const,
+                    highlights: [
+                        { hlId: 2, text: "h", virtText: "h" },
+                        { hlId: 2, text: "h", virtText: " " },
+                    ],
+                    line: 0,
+                    col: 1,
+                },
+            ],
+        },
+        {
+            testName: "redrawing a virtual highlight over a double-wide character redraws both cells",
+            events: [
+                {
+                    line: 0,
+                    vimCol: 0,
+                    vimCells: [["a", 2], ["1"]],
+                    lineText: "你",
+                    tabSize: 4,
+                },
+                {
+                    line: 0,
+                    vimCol: 1,
+                    vimCells: [["2", 2]],
+                    lineText: "你",
+                    tabSize: 4,
+                },
+            ],
+            expectedRanges: [
+                {
+                    textType: "virtual" as const,
+                    highlights: [
+                        { hlId: 2, text: "a", virtText: "a" },
+                        { hlId: 2, text: "2", virtText: "2" },
+                    ],
+                    line: 0,
+                    col: 0,
+                },
+            ],
+        },
+        {
+            testName: "two highlights on the same line and same hlId should produce two ranges",
+            events: [
+                {
+                    line: 0,
+                    vimCol: 0,
+                    vimCells: [["h", 2], ["e"], ["l"], ["l"], ["o"]],
+                    lineText: "hello world hello",
+                    tabSize: 4,
+                },
+                {
+                    line: 0,
+                    vimCol: 12,
+                    vimCells: [["h", 2], ["e"], ["l"], ["l"], ["o"]],
+                    lineText: "hello world hello",
+                    tabSize: 4,
+                },
+            ],
+            expectedRanges: [
+                {
+                    textType: "normal" as const,
+                    hlId: 2,
+                    line: 0,
+                    startCol: 0,
+                    endCol: 5,
+                },
+                {
+                    textType: "normal" as const,
+                    hlId: 2,
+                    line: 0,
+                    startCol: 12,
+                    endCol: 17,
+                },
+            ],
+        },
+    ] as {
+        testName: string;
+        events: {
+            line: number;
+            vimCol: number;
+            vimCells: VimCell[];
+            lineText: string;
+            tabSize: number;
+        }[];
+        expectedRanges: HighlightRange[];
+    }[];
+
+    testCases.forEach(({ testName, events, expectedRanges }) => {
```

Comment:
These tests don't use the spawned VSCode instance - why move them from unit tests?

Reply:
I understand your thinking. I also anticipated that this question might be raised here. I really don't think it's necessary to separate the logic just for testing purposes; testing it directly in VSCode is also very quick. Moreover, since this is a VSCode extension, there's no need to get bogged down with this issue.

Reply:
Understood. Frankly, on my own, I was always quite frustrated with the limitation we had of not being able to use VSCode imports.

I don't think we necessarily need to factor our code _just_ for testing, but I do think we should separate out tests that are manipulating inputs, and those that are calling code. They're ultimately very different kinds of tests, and if for no other reason other than organization, would be a good idea to do.